compiler/optimizer/LoopVersioner.cpp

```diff
@@ -7576,63 +7576,114 @@ void TR_LoopVersioner::collectAllExpressionsToBeChecked(List<TR::TreeTop> *nullC
+               {
+               // Currently it's not possible to use this class to eliminate
+               // the type check in AOT. That would require a verification
+               // record to check the subtyping relationship.
+               cl = fe()->getClassFromSignature(sig, len, symRef->getOwningMethod(comp()));
+               }
 
             int32_t otherLen;
             char *otherSig = otherSymRef->getOwningMethod(comp())->classNameOfFieldOrStatic(otherSymRef->getCPIndex(), otherLen);
             dumpOptDetails(comp(), "For node %p len %d other len %d sig %p other sig %p\n", node, len, otherLen, sig, otherSig);
-            if (otherSig)
-               otherClassObject = otherSymRef->getOwningMethod(comp())->getDeclaringClassFromFieldOrStatic(comp(),otherSymRef->getCPIndex());
+            TR::ResolvedMethodSymbol *owningMethodSym = otherSymRef->getOwningMethodSymbol(comp());
+            int32_t classCPI = 0;
+            int32_t fieldCPI = otherSymRef->getCPIndex();
+            if (otherSymRef->getSymbol()->isShadow() && fieldCPI >= 0)
+               {
+               TR_ResolvedMethod *owningMethod = owningMethodSym->getResolvedMethod();
+               classCPI = owningMethod->classCPIndexOfFieldOrStatic(fieldCPI);
+               bool aotOK = true;
+               otherClassObject = owningMethod->getClassFromConstantPool(comp(), classCPI, aotOK);
+               }
+#ifdef J9_PROJECT_SPECIFIC
+            else
+               {
+               switch (otherSymRef->getReferenceNumber() - comp()->getSymRefTab()->getNumHelperSymbols())
+                  {
+                  case TR::SymbolReferenceTable::classFromJavaLangClassSymbol:
+                  case TR::SymbolReferenceTable::classFromJavaLangClassAsPrimitiveSymbol:
+                     {
+                     bool aotOK = true;
+                     otherClassObject = comp()->getClassClassPointer(aotOK);
+                     classCPI = -1;
+                     if (otherClassObject == NULL)
+                        {
+                        // A type test against java/lang/Class is mandatory but
+                        // the class pointer is not forthcoming. At this point
+                        // it would be difficult to prevent the transformation
+                        // responsible for hoisting this load.
+                        comp()->failCompilation<TR::CompilationException>(
+                           "failed to find java/lang/Class during versioning");
+                        }
+                     break;
+                     }
+                  }
+               }
+#endif
 
-            //the call to findOrCreateClassSymbol is safe even though we pass CPI of -1 since the call above getClassFromSignature is not vetted for AOT
-            if (cl && otherClassObject && (fe()->isInstanceOf(cl, otherClassObject, true) == TR_yes))
+            if (otherClassObject == NULL)
```

Comment:
Why can we continue with the versioning in this case but not in the j/l/Class case ?

Reply:
The typical reason to have `otherClassObject == NULL` here would be that it wasn't set just above, because `otherSymRef` is either:

- an array shadow, which is okay because array accesses are handled elsewhere;
- the VFT, which is okay because all objects have it; or
- a field of a native struct, e.g. `classRomPtrSymbol`, in which case there is no way to generate a test.

In the last case, `firstChild` has to be such that, if it's safe to evaluate, then the result is always a `J9Class` (in this example), or more generally whatever it is that's expected by `node`. This is the same condition required for verifiability in (OpenJ9's) `transformIndirectLoadChain()`:

> For Java object references:
> - Verifiable if it is known to point at an object or null
> - Verified if it is non-null and points to an object that contains the field being loaded
>
> _For C structs pointers:_
> - _Verifiable if it is known to point to a struct of the proper type or null_
> - _Verified trivially by virtue of being verifiable and non null_

(emphasis mine)

But now that you mention it, it's also possible for `getClassFromConstantPool` to return null in AOT even though `isVettedForAOT=true`. I'll broaden the condition for bailing out to include circumstances where we should have gotten the class from the constant pool.

Reply:
Updated